// solargraph-utils/out/StdioProvider.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
const commands_1 = require("./commands");
class StdioProvider {
    constructor(configuration) {
        this.configuration = configuration;
        this._listening = false;
    }
    start() {
        return new Promise((resolve, reject) => {
            this.child = commands_1.solargraphCommand(['stdio'], this.configuration);
            let buffer = '';
            let that = this;
            this.child.stderr.on('data', (data) => {
                console.log(data.toString());
                if (!that.isListening()) {
                    that._listening = true;
                    buffer += data.toString();
                    var match = buffer.match(/PID=([0-9]*)/);
                    if (match) {
                        that._pid = parseInt(match[2]);
                        resolve();
                    }
                }
            });
            this.child.on('error', (err) => {
                if (!that.isListening()) {
                    reject(err);
                }
            });
            this.child.on('exit', (code) => {
                if (!that.isListening()) {
                    reject(buffer);
                }
            });
        });
    }
    stop() {
        if (!this.child) {
            console.warn('The server is not running.');
        }
        else {
            this.child.kill();
            if (this._pid) {
                process.kill(this._pid);
            }
            this._pid = null;
            this._listening = false;
        }
    }
    restart() {
        this.stop();
        return this.start();
    }
    isListening() {
        return this._listening;
    }
    get process() {
        return this.child;
    }
}
exports.StdioProvider = StdioProvider;
